File: minegame-database/src/main/java/top/nipuru/minegame/database/util/PostgreUtils.java
package top.nipuru.minegame.database.util;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class PostgreUtils {

    public static Object getObject(ResultSet rs, int index, Class<?> clazz) throws SQLException {
        if (clazz == int.class) {
            return rs.getInt(index);
        } else if (clazz == long.class) {
            return rs.getLong(index);
        } else if (clazz == float.class) {
            return rs.getFloat(index);
        } else if (clazz == double.class) {
            return rs.getDouble(index);
        } else if (clazz == boolean.class) {
            return rs.getBoolean(index);
        } else if (clazz == String.class) {
            return rs.getString(index);
        } else if (clazz == byte[].class) {
            return rs.getBytes(index);
        } else if (clazz == String[].class) {
            return rs.getArray(index).getArray();
        } else if (clazz == int[].class) {
            Integer[] array = (Integer[]) rs.getArray(index).getArray();
            int[] ints = new int[array.length];
            for (int i = 0; i < array.length; i++) {
                ints[i] = array[i];
            }
            return ints;
        } else if (clazz == long[].class) {
            Long[] array = (Long[]) rs.getArray(index).getArray();
            long[] longs = new long[array.length];
            for (int i = 0; i < array.length; i++) {
                longs[i] = array[i];
            }
            return longs;
        } else if (clazz == float[].class) {
            Float[] array = (Float[]) rs.getArray(index).getArray();
            float[] floats = new float[array.length];
            for (int i = 0; i < array.length; i++) {
                floats[i] = array[i];
            }
            return floats;
        } else if (clazz == double[].class) {
            Double[] array = (Double[]) rs.getArray(index).getArray();
            double[] doubles = new double[array.length];
            for (int i = 0; i < array.length; i++) {
                doubles[i] = array[i];
            }
            return doubles;
        } else if (clazz == boolean[].class) {
            Boolean[] array = (Boolean[]) rs.getArray(index).getArray();
            boolean[] booleans = new boolean[array.length];
            for (int i = 0; i < array.length; i++) {
                booleans[i] = array[i];
            }
            return booleans;
        }
        throw new IllegalArgumentException("Unsupported type: " + clazz);
    }

    public static void setObject(Connection con, PreparedStatement ps, int index, Object object) throws SQLException {
        Class<?> clazz = object.getClass();
        if (clazz == Integer.class) {
            ps.setInt(index, (int) object);
        } else if (clazz == Long.class) {
            ps.setLong(index, (long) object);
        } else if (clazz == Float.class) {
            ps.setFloat(index, (float) object);
        } else if (clazz == Double.class) {
            ps.setDouble(index, (double) object);
        } else if (clazz == Boolean.class) {
            ps.setBoolean(index, (boolean) object);
        } else if (clazz == String.class) {
            ps.setString(index, (String) object);
        } else if (clazz == byte[].class) {
            ps.setBytes(index, (byte[]) object);
        } else if (clazz == String[].class) {
            ps.setObject(index, object);
        } else if (clazz == int[].class) {
            int[] ints = (int[]) object;
            Integer[] data = new Integer[ints.length];
            for (int i = 0; i < ints.length; ++i) {
                data[i] = ints[i];
            }
            ps.setArray(index, con.createArrayOf("INTEGER", data));
        } else if (clazz == long[].class) {
            long[] longs = (long[]) object;
            Long[] data = new Long[longs.length];
            for (int i = 0; i < longs.length; ++i) {
                data[i] = longs[i];
            }
            ps.setArray(index, con.createArrayOf("BIGINT", data));
        } else if (clazz == float[].class) {
            float[] floats = (float[]) object;
            Float[] data = new Float[floats.length];
            for (int i = 0; i < floats.length; ++i) {
                data[i] = floats[i];
            }
            ps.setArray(index, con.createArrayOf("REAL", data));
        } else if (clazz == double[].class) {
            double[] doubles = (double[]) object;
            Double[] data = new Double[doubles.length];
            for (int i = 0; i < doubles.length; ++i) {
                data[i] = doubles[i];
            }
            ps.setArray(index, con.createArrayOf("DOUBLE PRECISION", data));
        } else if (clazz == boolean[].class) {
            boolean[] booleans = (boolean[]) object;
            Boolean[] data = new Boolean[booleans.length];
            for (int i = 0; i < booleans.length; ++i) {
                data[i] = booleans[i];
            }
            ps.setArray(index, con.createArrayOf("BOOLEAN", data));
        } else {
            throw new IllegalArgumentException("Unsupported type: " + clazz);
        }
    }

    public static String getSqlType(Class<?> fieldType) {
        if (fieldType == String.class) {
            return "TEXT";
        } else if (fieldType == Integer.class || fieldType == int.class) {
            return "INTEGER";
        } else if (fieldType == Long.class || fieldType == long.class) {
            return "BIGINT";
        } else if (fieldType == Boolean.class || fieldType == boolean.class) {
            return "BOOLEAN";
        } else if (fieldType == Double.class || fieldType == double.class) {
            return "DOUBLE PRECISION";
        } else if (fieldType == Float.class || fieldType == float.class) {
            return "REAL";
        } else if (fieldType == byte[].class) {
            return "BYTEA";
        } else if (fieldType.isArray()) {
            return getSqlType(fieldType.getComponentType()) + "[]";
        } else {
            throw new IllegalArgumentException("Unsupported type: " + fieldType.getSimpleName());
        }
    }

    public static String mapFieldName(String fieldName) {
        return "\"" + fieldName
                .replaceAll("([a-z])([A-Z])", "$1_$2")  // 在小写字母和大写字母之间添加下划线
                .toLowerCase() + "\"";
    }
}

File: minegame-database/src/main/kotlin/top/nipuru/minegame/database/util/PostgreUtils.kt
package top.nipuru.minegame.database.util

import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.*

@Suppress("UNCHECKED_CAST")
fun ResultSet.getObject(clazz: Class<*>, index: Int): Any {
    when (clazz) {
        Int::class.javaPrimitiveType -> {
            return this.getInt(index)
        }
        Long::class.javaPrimitiveType -> {
            return this.getLong(index)
        }
        Float::class.javaPrimitiveType -> {
            return this.getFloat(index)
        }
        Double::class.javaPrimitiveType -> {
            return this.getDouble(index)
        }
        Boolean::class.javaPrimitiveType -> {
            return this.getBoolean(index)
        }
        String::class.java -> {
            return this.getString(index)
        }
        ByteArray::class.java -> {
            return this.getBytes(index)
        }
        Array<String>::class.java -> {
            return this.getArray(index).array
        }
        IntArray::class.java -> {
            val array = this.getArray(index).array as Array<Int>
            val ints = IntArray(array.size)
            for (i in array.indices) {
                ints[i] = array[i]
            }
            return ints
        }
        LongArray::class.java -> {
            val array = this.getArray(index).array as Array<Long>
            val longs = LongArray(array.size)
            for (i in array.indices) {
                longs[i] = array[i]
            }
            return longs
        }
        FloatArray::class.java -> {
            val array = this.getArray(index).array as Array<Float>
            val floats = FloatArray(array.size)
            for (i in array.indices) {
                floats[i] = array[i]
            }
            return floats
        }
        DoubleArray::class.java -> {
            val array = this.getArray(index).array as Array<Double>
            val doubles = DoubleArray(array.size)
            for (i in array.indices) {
                doubles[i] = array[i]
            }
            return doubles
        }
        BooleanArray::class.java -> {
            val array = this.getArray(index).array as Array<Boolean>
            val booleans = BooleanArray(array.size)
            for (i in array.indices) {
                booleans[i] = array[i]
            }
            return booleans
        }
        else -> throw IllegalArgumentException("Unsupported type: $clazz")
    }
}


fun PreparedStatement.setObject(con: Connection, index: Int, obj: Any) {
    when (val clazz = obj.javaClass) {
        Int::class.java -> {
            this.setInt(index, obj as Int)
        }
        Long::class.java -> {
            this.setLong(index, obj as Long)
        }
        Float::class.java -> {
            this.setFloat(index, obj as Float)
        }
        Double::class.java -> {
            this.setDouble(index, obj as Double)
        }
        Boolean::class.java -> {
            this.setBoolean(index, obj as Boolean)
        }
        String::class.java -> {
            this.setString(index, obj as String)
        }
        ByteArray::class.java -> {
            this.setBytes(index, obj as ByteArray)
        }
        Array<String>::class.java -> {
            this.setObject(index, obj)
        }
        IntArray::class.java -> {
            val ints = obj as IntArray
            val data = arrayOfNulls<Int>(ints.size)
            for (i in ints.indices) {
                data[i] = ints[i]
            }
            this.setArray(index, con.createArrayOf("INTEGER", data))
        }
        LongArray::class.java -> {
            val longs = obj as LongArray
            val data = arrayOfNulls<Long>(longs.size)
            for (i in longs.indices) {
                data[i] = longs[i]
            }
            this.setArray(index, con.createArrayOf("BIGINT", data))
        }
        FloatArray::class.java -> {
            val floats = obj as FloatArray
            val data = arrayOfNulls<Float>(floats.size)
            for (i in floats.indices) {
                data[i] = floats[i]
            }
            this.setArray(index, con.createArrayOf("REAL", data))
        }
        DoubleArray::class.java -> {
            val doubles = obj as DoubleArray
            val data = arrayOfNulls<Double>(doubles.size)
            for (i in doubles.indices) {
                data[i] = doubles[i]
            }
            this.setArray(index, con.createArrayOf("DOUBLE PRECISION", data))
        }
        BooleanArray::class.java -> {
            val booleans = obj as BooleanArray
            val data = arrayOfNulls<Boolean>(booleans.size)
            for (i in booleans.indices) {
                data[i] = booleans[i]
            }
            this.setArray(index, con.createArrayOf("BOOLEAN", data))
        }
        else -> {
            throw IllegalArgumentException("Unsupported type: $clazz")
        }
    }
}

fun Class<*>.getSqlType(): String {
    return if (this == String::class.java) {
        "TEXT"
    } else if (this == Int::class.java || this == Int::class.javaPrimitiveType) {
        "INTEGER"
    } else if (this == Long::class.java || this == Long::class.javaPrimitiveType) {
        "BIGINT"
    } else if (this == Boolean::class.java || this == Boolean::class.javaPrimitiveType) {
        "BOOLEAN"
    } else if (this == Double::class.java || this == Double::class.javaPrimitiveType) {
        "DOUBLE PRECISION"
    } else if (this == Float::class.java || this == Float::class.javaPrimitiveType) {
        "REAL"
    } else if (this == ByteArray::class.java) {
        "BYTEA"
    } else if (this.isArray) {
        this.componentType.getSqlType() + "[]"
    } else {
        throw IllegalArgumentException("Unsupported type: " + this.simpleName)
    }
}

fun String.getSqlName(): String {
    return "\"" + this
        .replace("([a-z])([A-Z])".toRegex(), "$1_$2") // 在小写字母和大写字母之间添加下划线
        .lowercase(Locale.getDefault()) + "\""
}
